### PR TITLE
Add option to specify the allowed "php" wrapper types 

### PR DESCRIPTION
**Pull request by cgzones**
In addition of the current possibility to filter wrappers by their protocol name, also add the option to filter the **"php"** wrapper by the requested kind.
Especially the `filter` backend can be disabled that way.

See https://www.php.net/manual/en/wrappers.php.php and https://github.com/ambionics/cnext-exploits.

### CODE REVIEW

File: doc/source/config.rst [hunk 1]
```diff
@@ -395,6 +395,17 @@ to explicitly whitelist some `stream wrappers <https://secure.php.net/manual/en/
   sp.wrappers_whitelist.list("file,php,phar");
 
 
+Allowlist of the php stream-wrapper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:ref:`The php-stream-wrapper allowlist <php-stream-wrapper-allowlist-feature>`
+allows to explicitly allow the builtin `php streams <https://www.php.net/manual/en/wrappers.php.php>`__.
```
Comment (jvoisin) [inline doc/source/config.rst:402]: "the allowlist allows to allow" is a tad redundant :P

File: config/default_php8.rules [hunk 2]
```diff
@@ -26,6 +26,8 @@ sp.xxe_protection.enable();
 # PHP has a lot of wrappers, most of them aren't usually useful, you should
 # only enable the ones you're using.
 # sp.wrappers_whitelist.list("file,php,phar");
+# The "php" wrapper can be further filtered
```
Comment (jvoisin) [inline config/default_php8.rules:29]:
```suggestion
# The "php" wrapper can be further filtered, eg. to only allow `php://stdout`, `php://stdin` and `php://stderr`:
```

File: doc/source/features.rst [hunk 3]
```diff
@@ -397,6 +397,20 @@ Examples of related vulnerabilities
...
+
+The builtin `"php" stream wrapper <https://www.php.net/manual/en/wrappers.php.php>`__
+has support for common streams, like ``stdin``, ``stdout`` or ``stderr``, but
+also for the dangerous ``filter`` one.
+
+Examples of related vulnerability
+"""""""""""""""""""""""""""""""""
+
+- `CNEXT exploits <https://github.com/ambionics/cnext-exploits/>`__
+
```
Comment (jvoisin) [inline doc/source/features.rst:413]:
```suggestion
Examples of related vulnerabilities
"""""""""""""""""""""""""""""""""""

- `CNEXT exploits <https://github.com/ambionics/cnext-exploits/>`__
- Synacktiv's `php_filter_chain_generator <https://github.com/synacktiv/php_filter_chain_generator>`__ tool
- Ambionic's `wrapwrap <https://github.com/ambionics/wrapwrap>`__ tool
```

**Comment by jvoisin** [comment]
Wow, that's a pretty neat PR, thanks!

**Comment by cgzones** [comment]
In addition when the `filter` stream for the `php` wrapper is required, one could limit the depth of chained filters, e.g. set a value of *4* to permit `php:://filter/a/b/c/d/resource=foo` but deny `php:://filter/a/b/c/d/e/resource=foo`.

**Comment by jvoisin** [comment]
I'm not sure limiting the depth will provide meaningful security improvements, nor that it'll be used to be honest :/

**Comment by jvoisin** [comment]
This PR looks good to me, anything specific you're waiting on to mark it "Ready for review"?

**Comment by jvoisin** [comment]
Merged, thanks <3

**Pull request closed without merging**